src/arc_dst.rs

```diff
@@ -0,0 +1,476 @@
+where
+    T: Copy + Send + Sync + Hash + Eq + 'static,
+{
+    fn from(slice: &[T]) -> Self {
+        ArcIntern::<[T]>::new_with_copyable_init_val(slice, |slice| {
+            RefCount::<[T]>::from_slice(slice)
+        })
+    }
+}
+impl_from! { [T: Copy + Send + Sync + Hash + Eq + 'static] Vec<T>, ArcIntern<[T]> }
+impl_from! { [T: Copy + Send + Sync + Hash + Eq + 'static] Box<[T]>, ArcIntern<[T]> }
+impl_from! { [T: Copy + Send + Sync + Hash + Eq + 'static] Arc<[T]>, ArcIntern<[T]> }
+impl_from! { [T: Copy + Send + Sync + Hash + Eq + 'static] std::rc::Rc<[T]>, ArcIntern<[T]> }
+
+impl Default for ArcIntern<str> {
+    fn default() -> Self {
+        Self::from("")
+    }
+}
+impl<T> Default for ArcIntern<[T]>
+where
+    T: Copy + Send + Sync + Hash + Eq + 'static,
+{
+    fn default() -> Self {
+        Self::from(&[][..])
+    }
+}
+
+// implement some useful equal comparisons
+macro_rules! impl_eq {
+    ([$($vars:tt)*] $lhs:ty, $rhs: ty) => {
+        #[allow(unused_lifetimes)]
+        impl<'a, $($vars)*> PartialEq<$rhs> for $lhs {
+            #[inline]
+            fn eq(&self, other: &$rhs) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+            #[inline]
+            fn ne(&self, other: &$rhs) -> bool {
+                PartialEq::ne(&self[..], &other[..])
+            }
+        }
+
+        #[allow(unused_lifetimes)]
+        impl<'a, $($vars)*> PartialEq<$lhs> for $rhs {
+            #[inline]
+            fn eq(&self, other: &$lhs) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+            #[inline]
+            fn ne(&self, other: &$lhs) -> bool {
+                PartialEq::ne(&self[..], &other[..])
+            }
+        }
+    };
+}
+impl_eq! { [] ArcIntern<str>, str }
+impl_eq! { [] ArcIntern<str>, &'a str }
+impl_eq! { [] ArcIntern<str>, String }
+impl_eq! { [] ArcIntern<str>, std::borrow::Cow<'a, str> }
+impl_eq! { [] ArcIntern<str>, Box<str> }
+impl_eq! { [] ArcIntern<str>, std::rc::Rc<str> }
+impl_eq! { [] ArcIntern<str>, std::sync::Arc<str> }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, Vec<T> }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, [T] }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, &'a [T] }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, &'a mut [T] }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, std::borrow::Cow<'a, [T]> }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, Box<[T]> }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, std::rc::Rc<[T]> }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static] ArcIntern<[T]>, std::sync::Arc<[T]> }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static, const N: usize] ArcIntern<[T]>, [T; N] }
+impl_eq! { [T: Copy + Send + Sync + Hash + Eq + 'static, const N: usize] ArcIntern<[T]>, &[T; N] }
```

Comment:
I'm not at all clear why you would need all these constraints.
```suggestion
impl_eq! { [T: Eq ] ArcIntern<[T]>, Vec<T> }
impl_eq! { [T:  Eq] ArcIntern<[T]>, [T] }
impl_eq! { [T: Eq ] ArcIntern<[T]>, &'a [T] }
impl_eq! { [T: Eq ] ArcIntern<[T]>, &'a mut [T] }
impl_eq! { [T: Eq] ArcIntern<[T]>, std::borrow::Cow<'a, [T]> }
impl_eq! { [T: Eq] ArcIntern<[T]>, Box<[T]> }
impl_eq! { [T: Eq] ArcIntern<[T]>, std::rc::Rc<[T]> }
impl_eq! { [T: Eq] ArcIntern<[T]>, std::sync::Arc<[T]> }
impl_eq! { [T: Eq, const N: usize] ArcIntern<[T]>, [T; N] }
impl_eq! { [T: Eq ,const N: usize] ArcIntern<[T]>, &[T; N] }
```
I know you need the other constraints to create the arcintern in the first place but it's cleaner to only put the constraints needed for these impls on the impls.